### PR TITLE
Implement CPU and Memory runtime controls

### PR DESCRIPTION
**Pull request by PineappleIOnic**
## What does this PR do?

This PR implements CPU and Memory controls for Runtimes, adding new `cpus` and `memory` attributes onto the functions document and it's relevant API endpoints. `_APP_FUNCTIONS_CPUS` and `_APP_FUNCTIONS_MEMORY` are now utilised as the max amount of memory and CPUs that can be allocated to a function.

### To be implemented
- [x] Plan Differentiators in API Parameters
- [x] Using `_APP_FUNCTIONS_CPUS` and `_APP_FUNCTIONS_MEMORY` as limiters and to reflect that in the API
- [x] Implementing the API in console
- [x] Fixing the issue we're having with TS Enums having numeric names
- [x] Implement Tests

## Test Plan
1. Create a function and execution then check it's stats
2. Change the runtime memory, rebuild and check it's stats again
3. Change the runtime CPU, rebuild and check it's stats one more time

## Related PRs and Issues

- https://github.com/appwrite/console/pull/1138

## Checklist

- [X] Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?
- [ ] If the PR includes a change to an API's metadata (desc, label, params, etc.), does it also include updated API specs and example docs?

### CODE REVIEW

File: .env [hunk 1]
```diff
@@ -66,8 +66,8 @@ _APP_STORAGE_PREVIEW_LIMIT=20000000
 _APP_FUNCTIONS_SIZE_LIMIT=30000000
 _APP_FUNCTIONS_TIMEOUT=900
 _APP_FUNCTIONS_BUILD_TIMEOUT=900
-_APP_FUNCTIONS_CPUS=1
-_APP_FUNCTIONS_MEMORY=1024
```
Comment (Meldiron) [inline .env:70]: Lets use previous values again

File: app/controllers/api/functions.php [hunk 2]
```diff
@@ -160,6 +169,8 @@
     ->param('templateOwner', '', new Text(128, 0), 'The name of the owner of the template.', true)
     ->param('templateRootDirectory', '', new Text(128, 0), 'Path to function code in the template repo.', true)
     ->param('templateBranch', '', new Text(128, 0), 'Production branch for the repo linked to the function template.', true)
+    ->param('memory', 512, fn (array $plan) => empty($plan['memoryLimit']) ? new WhiteList($filterBelowThreshold(MEMORY_VALUES, System::getEnv('_APP_FUNCTIONS_MEMORY')), false, WhiteList::TYPE_INTEGER) : new WhiteList($filterBelowThreshold(MEMORY_VALUES, $plan['memoryLimit']), false, WhiteList::TYPE_INTEGER), 'Memory in MB allocated for the function.', true, ['plan'])
```
Comment (Meldiron) [inline app/controllers/api/functions.php:172]: I dont think $plan will be here,  it's Cloud-only.
I would suggest special Memory and CPU validators here instead. Then, on Cloud, we can override those validators and add billing logic.

File: app/controllers/api/functions.php [hunk 3]
```diff
@@ -442,6 +455,26 @@
...
     });
 
+App::get('/v1/functions/specs')
+    ->groups(['api', 'functions'])
+    ->desc('Get available function specs')
+    ->label('scope', 'functions.read')
+    ->label('sdk.auth', [APP_AUTH_TYPE_KEY])
+    ->label('sdk.namespace', 'functions')
+    ->label('sdk.method', 'getSpecs')
+    ->label('sdk.description', '/docs/references/functions/get-specs.md')
```
Comment (Meldiron) [inline app/controllers/api/functions.php:465]: Let's write markdown file

File: app/controllers/api/functions.php (hunk 2 above)
Comment (Meldiron) [reply app/controllers/api/functions.php:172]: This comment applies to all places where you used $plan

File: src/Appwrite/Platform/Workers/Builds.php [hunk 4]
```diff
@@ -410,6 +410,8 @@ protected function buildDeployment(Device $deviceForFunctions, Func $queueForFun
                             source: $source,
                             image: $runtime['image'],
                             version: $version,
+                            cpus: $function->getAttribute('cpus', 1),
+                            memory: $function->getAttribute('memory', 128),
```
Comment (Meldiron) [inline src/Appwrite/Platform/Workers/Builds.php:414]: Execution specs and build specs should remain different. We should either hard-code build specs (very high), or make it configurable separatelly.

File: src/Appwrite/Utopia/Response/Model/Func.php [hunk 5]
```diff
@@ -145,6 +145,18 @@ public function __construct()
...
                 'example' => false,
             ])
+            ->addRule('memory', [
+                'type' => self::TYPE_INTEGER,
+                'description' => 'Function execution memory limit in MB.',
+                'default' => 512,
+                'example' => 1024,
+            ])
+            ->addRule('cpus', [
+                'type' => self::TYPE_INTEGER,
```
Comment (Meldiron) [inline src/Appwrite/Utopia/Response/Model/Func.php:155]: Let's confirm with team lead.. Should this be integer, or string? If it's integer I would expect slider on Console. If it was string, I would expect select dropdown.
Same for developers using API, integer indicates I might be able to set it to anything. Using ENUM here might work better

File: src/Appwrite/Utopia/Response/Model/Func.php [hunk 6]
```diff
@@ -145,6 +145,18 @@ public function __construct()
...
+                'description' => 'Function execution cores.',
```
Comment (Meldiron) [inline src/Appwrite/Utopia/Response/Model/Func.php:156]:
```suggestion
                'description' => 'Function execution CPU cores.',
```

File: src/Executor/Executor.php [hunk 7]
```diff
@@ -83,8 +85,8 @@ public function createRuntime(
             'variables' => $variables,
             'remove' => $remove,
             'command' => $command,
-            'cpus' => $this->cpus,
-            'memory' => $this->memory,
```
Comment (Meldiron) [inline src/Executor/Executor.php:87]: We can remove those $this variables, they shouldn' be needed anymore

File: src/Appwrite/Utopia/Response/Model/Func.php (hunk 5 above)
Comment (PineappleIOnic) [reply src/Appwrite/Utopia/Response/Model/Func.php:155]: Enum will probably make more sense here, will update appropiately

File: src/Appwrite/Platform/Workers/Builds.php (hunk 4 above)
Comment (PineappleIOnic) [reply src/Appwrite/Platform/Workers/Builds.php:414]: I hard-coded 2 cpu cores and 1024 memory, that should be enough right?

File: src/Appwrite/Utopia/Response/Model/Func.php (hunk 5 above)
Comment (PineappleIOnic) [reply src/Appwrite/Utopia/Response/Model/Func.php:155]: Can we add enums to models? I can't seem to see a `self::TYPE_ENUM` in the Model class

File: app/controllers/api/functions.php [hunk 8]
```diff
@@ -53,6 +53,9 @@
 
 include_once __DIR__ . '/../shared/api.php';
 
+const MEMORY_VALUES = [512, 1024, 2048, 4096, 8192, 16384];
+const CPU_VALUES = [1, 2, 4, 8, 16];
```
[the remark this answers is not in the document]
Comment (PineappleIOnic) [reply app/controllers/api/functions.php]: I moved these into the new validators I just added

File: app/controllers/api/functions.php (hunk 2 above)
Comment (PineappleIOnic) [reply app/controllers/api/functions.php]: Created the new validators, please re-review

File: app/controllers/api/functions.php (hunk 3 above)
Comment (PineappleIOnic) [reply app/controllers/api/functions.php:458]: Added 👍

File: src/Executor/Executor.php (hunk 7 above)
Comment (PineappleIOnic) [reply src/Executor/Executor.php:87]: Alright, done

File: docs/references/functions/get-specs.md [hunk 9]
```diff
@@ -0,0 +1 @@
+Get allowed function and cpus values on this instance.
```
Comment (Meldiron) [inline docs/references/functions/get-specs.md:1]:
```suggestion
Get allowed function memory and cpus values on this instance.
```

File: src/Appwrite/Utopia/Response/Model/Func.php (hunk 5 above)
Comment (Meldiron) [reply src/Appwrite/Utopia/Response/Model/Func.php:155]: Not sure, if we never did it, we can do STRING. Check what we did for `plan` in billing, I think that behaves as enum

File: src/Executor/Executor.php (hunk 7 above)
Comment (Meldiron) [reply src/Executor/Executor.php:87]: I meant those:

CleanShot 2024-07-10 at 09 16 16@2x

File: src/Executor/Executor.php [hunk 10]
```diff
@@ -65,6 +65,8 @@ public function createRuntime(
         string $source,
         string $image,
         string $version,
+        int $cpus = 1,
+        int $memory = 512,
```
Comment (Meldiron) [inline src/Executor/Executor.php:69]: I dont see those two being used.

File: src/Executor/Executor.php (hunk 7 above)
Comment (PineappleIOnic) [reply src/Executor/Executor.php:87]: Whoops, my bad. Fixing now

File: src/Executor/Executor.php (hunk 10 above)
Comment (PineappleIOnic) [reply src/Executor/Executor.php:69]: Yeah since I accidentally deleted the part that used it, adding it back now

**Comment by PineappleIOnic** [comment]
Closing this to retarget it to 1.5.x

**Pull request closed without merging**